### PR TITLE
SUBMARINE-559. [API] Define Swagger API for pre-defined template registration/delete, etc.

### PR DESCRIPTION
**Pull request by JohnTing**
### What is this PR for?

Make basic rest api for management experiment template.

get, post, del, patch
http://localhost/V1/template

### What type of PR is it?
[Bug Fix | Improvement | Feature | Documentation | Hot Fix | Refactoring]

### Todos
* [x] - REST API
* [x] - test
* [x] - e2e test
* [x] - parameter Mapping

### What is the Jira issue?
https://issues.apache.org/jira/projects/SUBMARINE/issues/SUBMARINE-559

### How should this be tested?
* First time? Setup Travis CI as described on https://submarine.apache.org/contribution/contributions.html#continuous-integration
* Strongly recommended: add automated unit tests for any new or changed behavior
* Outline any manual steps to test the PR here.

### Screenshots (if appropriate)
image
image
image
image

image
image

### Questions:
* Does the licenses files need update? Yes/No
* Is there breaking changes for older versions? Yes/No
* Does this needs documentation? Yes/No

### CODE REVIEW

**Comment by pingsutw** [comment]
Could you take a look at test failure?
https://travis-ci.org/github/apache/submarine/jobs/711883279
```bash
Results :
Tests in error: 
  ExperimentTemplateServiceTest.testSelect:61->compareTemplates:115 ClassCast co...
  ExperimentTemplateServiceTest.testUpdate:87->compareTemplates:115 ClassCast co...
```

File: submarine-server/server-core/src/test/java/org/apache/submarine/server/workbench/database/service/ExperimentTemplateServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    boolean deleteResult = templateService.deleteById(templateDb2.getId());
+    assertTrue(deleteResult);
+  }
+
+  private void compareTemplates(ExperimentTemplate template, ExperimentTemplate templateDb) {
+    assertEquals(template.getAuthor(), templateDb.getAuthor());
+    assertEquals(template.getName(), templateDb.getName());
+    assertEquals(template.getParameters().size(), templateDb.getParameters().size());
+    
+    if (template.getParameters().size() == templateDb.getParameters().size()) {
```
Comment (pingsutw) [inline submarine-server/server-core/src/test/java/org/apache/submarine/server/workbench/database/service/ExperimentTemplateServiceTest.java:111]: this line could be remove

File: docs/database/submarine.sql [hunk 2]
```diff
@@ -264,3 +264,20 @@ CREATE TABLE `params` (
   `job_name` varchar(32) NOT NULL COMMENT 'JOB NAME to which this parameter belongs to: Part of *Primary Key* for ``params`` table.',
   PRIMARY KEY  (`id`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
+-- ----------------------------
+-- Table structure for templates
+-- ----------------------------
+DROP TABLE IF EXISTS `templates`;
+CREATE TABLE `templates` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
```
Comment (pingsutw) [inline docs/database/submarine.sql:273]: change to `VARCHAR`. e.g. `templates_12345678`
https://github.com/apache/submarine/blob/master/submarine-server/server-api/src/main/java/org/apache/submarine/server/api/environment/EnvironmentId.java

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                  @ApiResponse(description = "successful operation", content = @Content(
+                          schema = @Schema(implementation = JsonResponse.class)))})
+
+  public Response postTemplate(ExperimentTemplate template) {
+    boolean result = false;
+    try {
+      result = templateService.insert(template);
+    } catch (Exception e) {
+      LOG.error(e.toString());
+      return new JsonResponse.Builder<Boolean>(Response.Status.OK).success(false).build();
```
Comment (pingsutw) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java:102]:
```suggestion
      return new JsonResponse.Builder<Boolean>(Response.Status.BAD_REQUEST).success(false).build();
```

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+  public Response getTemplate(@PathParam("id") String id) {
+    ExperimentTemplate template;
+    try {
+      template = new ExperimentTemplate();
+      template.setId(id);
+      template = templateService.selectByPrimaryKeySelective(template).get(0);
+    } catch (Exception e) {
+      LOG.error(e.toString());
+      return new JsonResponse.Builder<Boolean>(Response.Status.OK).success(false).build();
```
Comment (pingsutw) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java:123]:
```suggestion
      return new JsonResponse.Builder<Boolean>(Response.Status.BAD_REQUEST).success(false).build();
```

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                  @ApiResponse(description = "successful operation", content = @Content(
+                          schema = @Schema(implementation = JsonResponse.class)))})
+
+  public Response deleteTemplate(@PathParam("id") String id) {
+    boolean result = false;
+    try {
+      result = templateService.deleteById(id);
+    } catch (Exception e) {
+      LOG.error(e.toString());
+      return new JsonResponse.Builder<Boolean>(Response.Status.OK).success(false).build();
```
Comment (pingsutw) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java:143]:
```suggestion
      return new JsonResponse.Builder<Boolean>(Response.Status.BAD_REQUEST).success(false).build();
```

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                          schema = @Schema(implementation = JsonResponse.class)))})
+
+  public Response putTemplate(@PathParam("id") String id, ExperimentTemplate template) {
+    template.setId(id);
+    boolean result = false;
+    try {
+      result = templateService.update(template);
+    } catch (Exception e) {
+      LOG.error(e.toString());
+      return new JsonResponse.Builder<Boolean>(Response.Status.OK).success(false).build();
```
Comment (pingsutw) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java:163]:
```suggestion
      return new JsonResponse.Builder<Boolean>(Response.Status.BAD_REQUEST).success(false).build();
```

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  @POST
+  @Operation(summary = "post template",
+          tags = {"template"},
+          responses = {
+                  @ApiResponse(description = "successful operation", content = @Content(
+                          schema = @Schema(implementation = JsonResponse.class)))})
+
+  public Response postTemplate(ExperimentTemplate template) {
```
Comment (pingsutw) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java:96]:
```suggestion
  public Response createTemplate(ExperimentTemplate template) {
```

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/rest/ExperimentTemplateRestApi.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+import io.swagger.v3.oas.annotations.Operation;
+import io.swagger.v3.oas.annotations.media.Content;
+import io.swagger.v3.oas.annotations.media.Schema;
+import io.swagger.v3.oas.annotations.responses.ApiResponse;
+
+/**
+ * ExperimentTemplate REST API v1. It can accept {@link ExperimentTemplateSpec} to create a
+ * experimentTemplate.
+ */
+@Path(RestConstants.V1 + "/" + RestConstants.ENVIRONMENTS)
```
Comment (wangdatan) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/rest/ExperimentTemplateRestApi.java:51]: Why this is environment?

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/utils/JsonTypeHandler.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.sql.CallableStatement;
+import java.sql.PreparedStatement;
+
+import org.apache.ibatis.type.BaseTypeHandler;
+import org.apache.ibatis.type.JdbcType;
+
+import com.google.gson.Gson;
+
+public class JsonTypeHandler<T> extends BaseTypeHandler<T> {
```
Comment (wangdatan) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/utils/JsonTypeHandler.java:33]: Is it required? (I didn't see it being used)

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/utils/ListTypeHandler.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.sql.CallableStatement;
+import java.sql.PreparedStatement;
+
+import org.apache.ibatis.type.BaseTypeHandler;
+import org.apache.ibatis.type.JdbcType;
+
+import com.google.gson.Gson;
+
+public class ListTypeHandler<T extends Object> extends BaseTypeHandler<List<T>> {
```
Comment (wangdatan) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/utils/ListTypeHandler.java:34]: Is it required? (I didn't see it being used)

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/utils/JsonTypeHandler.java (hunk 9 above)
Comment (JohnTing) [reply submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/utils/JsonTypeHandler.java]: This was used in the previous commit.
But this one is not used now.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/utils/ListTypeHandler.java (hunk 10 above)
Comment (JohnTing) [reply submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/database/utils/ListTypeHandler.java]: This was used in the previous commit.
But this one is not used now.

File: submarine-server/server-core/src/test/java/org/apache/submarine/server/workbench/database/service/ExperimentTemplateServiceTest.java (hunk 1 above)
Comment (JohnTing) [reply submarine-server/server-core/src/test/java/org/apache/submarine/server/workbench/database/service/ExperimentTemplateServiceTest.java]: deleted

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java (hunk 3 above)
Comment (JohnTing) [reply submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java]: Fixed

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java (hunk 5 above)
Comment (JohnTing) [reply submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java]: Fixed

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java (hunk 7 above)
Comment (JohnTing) [reply submarine-server/server-core/src/main/java/org/apache/submarine/server/workbench/rest/ExperimentTemplateRestApi.java]: Fixed

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/rest/ExperimentTemplateRestApi.java (hunk 8 above)
Comment (JohnTing) [reply submarine-server/server-core/src/main/java/org/apache/submarine/server/rest/ExperimentTemplateRestApi.java]: Fixed

**Comment by wangdatan** [comment]
Thanks @JohnTing , it looks better to me. One question: Does our ExperimentSpec support `{{...}}` format? (In other words, is it a valid YAML syntax?) 

Also, we need to do validations for the parameters specified in the experiment spec. For parameters in `{{..}}`, and In ParameterSpec, we need make sure they have 1-1 mapping.

I think replace parameter logic will be done when we implement submit pre-defined template logics.

**Comment by JohnTing** [comment]
@wangdatan , thanks for your comment,

I will add replace parameter logic and modify the replacement format.

**Comment by wangdatan** [comment]
> @wangdatan , thanks for your comment,
> 
> I will add replace parameter logic and modify the replacement format.

Thanks @JohnTing , I think modify/replace logic can be done in the submit/execution JIRA: https://issues.apache.org/jira/browse/SUBMARINE-558 

We need to make sure verification/validation can be done in this JIRA. (Make sure 1-1 match of parameterSpec and experimentSpec).

**Comment by tangzhankun** [comment]
@JohnTing , thanks for the great efforts!
@jiwq I'm thinking once this experiment template is ready, will it change the run experiment spec implementation dramatically?

**Comment by liuxunorg** [comment]
@JohnTing Thank you contribution this feature.
Will merge if no more comments.

**Pull request closed without merging**